server/storage/mvcc/watchable_store.go

```diff
@@ -403,11 +404,39 @@ func (s *watchableStore) syncWatchers() int {
 	}
 	slowWatcherGauge.Set(float64(s.unsynced.size() + vsz))
 
-	return s.unsynced.size()
+	return s.unsynced.size(), evs
+}
+
+// rangeEventsWithReuse returns events in range [minRev, maxRev), while reusing already provided events.
+func rangeEventsWithReuse(lg *zap.Logger, b backend.Backend, evs []mvccpb.Event, minRev, maxRev int64) []mvccpb.Event {
+	if len(evs) == 0 {
+		return rangeEvents(lg, b, minRev, maxRev)
+	}
+	if evs[0].Kv.ModRevision > minRev {
+		evs = append(rangeEvents(lg, b, minRev, evs[0].Kv.ModRevision), evs...)
+	}
+	prefixIndex := 0
+	for prefixIndex < len(evs) && evs[prefixIndex].Kv.ModRevision < minRev {
+		prefixIndex++
+	}
+	evs = evs[prefixIndex:]
+
+	if len(evs) == 0 {
+		return rangeEvents(lg, b, minRev, maxRev)
+	}
+	if evs[len(evs)-1].Kv.ModRevision+1 < maxRev {
+		evs = append(evs, rangeEvents(lg, b, evs[len(evs)-1].Kv.ModRevision+1, maxRev)...)
+	}
+	suffixIndex := len(evs) - 1
+	for suffixIndex >= 0 && evs[suffixIndex].Kv.ModRevision >= maxRev {
```

Comment:
In which case `evs[suffixIndex].Kv.ModRevision >= maxRev` can be `true`?

Reply:
From the calls in `syncWatchers` this should never happen as it would mean that curRev decreased, which should not be possible. However when writing the function I decided implement the function to handle this case for testing. 

It allows us to use the same tests for both `rangeEvents` and `rangeEventsWithReuse`, it's a very nice property that reduces the chance of error in test.

Reply:
> that reduces the chance of error in test.

In that case, you need to fix the test instead of the production code.

> I decided implement the function to handle this case for testing.

On the contrary, I think it would be better to raise an error or even a panic in such case, because it means a major or critical bug in production.

Reply:
Not every low level function needs to be aware of every assumption. Don't know if this rule has a name, but "high level functions should assume less and require more, however the lower the call stack you go you can assume more and require less". For me definition of function "rageRevisions" should handle arbitrary range. 

> In that case, you need to fix the test instead of the production code.

It's not issue with test. Tests that should be easy to understand and maintainable, this makes the code easier to test.

> > I decided implement the function to handle this case for testing.
>
> On the contrary, I think it would be better to raise an error or even a panic in such case, because it means a major or critical bug in production.

It's not responsibility of function for reading revisions to validate if the revision in the etcd is non-decreasing. If we couldn't implement this case then we should add a assertion to protect function returning bad response, however implementation is simple enough that we can handle it.

Even though we expect global revision to never decrease, don't know if this is necessarily true in watch serving code, especially during restore. We track revision progress per watcher, so if revision ever decreased we would just skip sending event. I think this case was exactly confirmed in https://github.com/etcd-io/etcd/issues/15271.

  Don't think we should introduce this assertion now.

Reply:
Raised https://github.com/etcd-io/etcd/pull/18980 to verify this in a higher level function.